### PR TITLE
Cleanup throws in irep (de)serialization

### PR DESCRIPTION
**Pull request by hannes-steffenhagen-diffblue**

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: c8d7718b6f5432ceeedf50f90b3bccb1a20f2889).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84845938

File: src/util/irep_serialization.cpp [hunk 1]
```diff
@@ -95,8 +96,7 @@ void irep_serializationt::read_irep(
 
   if(in.get()!=0)
   {
-    std::cerr << "irep not terminated\n";
-    throw 0;
+    throw deserialization_exceptiont("irep not terminated");
```
Comment (peterschrammel) [inline src/util/irep_serialization.cpp:99]: Where is this caught?

Comment (hannes-steffenhagen-diffblue) [reply src/util/irep_serialization.cpp:99]: Good catch (no pun intended) I'd forgotten about that (I originally introduced something like a `cprover_exception_baset` and caught _that_, but then decided against it as I didn't find it saved that much effort compared to not having it...)

File: src/util/exception_utils.h [hunk 2]
```diff
@@ -33,4 +33,23 @@ class invalid_user_input_exceptiont
   std::string what() const noexcept;
 };
 
+class system_exceptiont
```
Comment (tautschnig) [inline src/util/exception_utils.h]: Please add comments: when should this be used, i.e., what is (or isn't) a "system" exception.

File: src/util/exception_utils.h [hunk 3]
```diff
@@ -52,4 +52,13 @@ class system_exceptiont
   }
 };
 
+class deserialization_exceptiont
```
Comment (tautschnig) [inline src/util/exception_utils.h]: See other PR: this should likely derive from some common invalid-input exception base class.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: b04c67c0d1ad984b33a566fc0fa2367f4d26d7c5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85131105
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 54d54872739c9785d3e80a98be26fcae8cb0a7d1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85132310

File: src/util/irep_serialization.cpp (hunk 1 above)
Comment (hannes-steffenhagen-diffblue) [reply src/util/irep_serialization.cpp:99]: @peterschrammel This is now sorted with #2996

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 9373209777f9b0793cdcc77dab84c3fd498e582c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85576436

**Review by tautschnig — APPROVED** [review]

**Pull request merged**